ls1intum/Artemis
# Feature/New Xcode Template using a Fastlane BuildPlan

### Checklist

### Motivation and Context


This PR replaces the current Xcode template. 
The new template has a different structure: it now uses a xcworkspace (in the test repository) containing two projects, one for the tests and one for the exercise / solution. The BuildPlans now work with fastlane instead of a xcodebuild command. 
The new template gives students the possibility to create new files in the project and after cloning the repository they are able to open the xcodeproj file in Xcode to work on the project, build it and test it on an iPhone Simulator. 

The template gets extended in a Follow-Up PR. It then should include the strategy pattern with sorting as the other templates do. The UI component should allow selecting a sorting strategy, generate a list of random dates via a simple button tap, sort them using another button and display the (un-) sorted list.

### Description

The template structure as well as the BuildPlans changed completely. The changes are described above and below.
Also I changed the name of the text field to enter a package name while creating a Swift programming exercise. Now it is called App Name, this is used for the dynamic naming of the files.

### Steps for Testing


1. You need to use Test Server 1 (this is necessary because of Swift)
2. Log in to Artemis
2. Navigate to Course Administration
3. Create a programming exercise and select `Swift` as programming language and `Xcode` as project type (there should be an option to type in an `App Name`) or use the following exercise: https://artemistest.ase.in.tum.de/course-management/480/programming-exercises/10033 
4. Check that the solution result is `2 of 2 passed` and the exercise result is `0 of 2 passed`
5. Clone the three repositories (for the next steps you need Xcode installed on your machine)
    5.1 In the test repository: Open the `xcworkspace` and see the two linked projects, and the `Unit` and `UI` `tests` in the test projects; also check that a `fastlane` folder is present in the repository containing the necessary configuration files (`Appfile`, `Fastfile`, `Scanfile` and a `readme`)
    5.2 In the exercise repository: Open the `xcodeproj` and see the exercise template -> solve the tasks and push your solution
    5.3 In the solution repository: Open the `xcodeproj` and check the solution 
6. See how the result of the exercise gets updated
7. Check the BuildPlans on Bamboo
    7.1 They should have two script tasks: build -> calls `fastlane build`; test -> calls `fastlane test` and `exit 0` (this is necessary due to a warning which would cause the build to fail even if it should not fail)
   7.2 The Junit Parser should read the `**/report.junit` file
   7.3 The requirements should be: `Simulator - iOS 14.5` and `fastlane`


### Review Progress





- Code Review
- Manual Tests


### Screenshots


Here you can see the textfield for the App Name:
<img width="573" alt="App Name" src="https://user-images.githubusercontent.com/82283581/132768521-71ed4817-9c47-4295-9880-87eef7742a19.png">

This is the testing step regarding the BuildPlan Configuration. In the video it did it with the BASE BuildPlan but the SOLUTION BuildPlan shows exactly the same:
https://user-images.githubusercontent.com/82283581/134001595-aadc9819-2d9b-4ffc-8e0c-68acca5d1282.mov
